### ISSUE TITLE
Performance issue with --slice-formula

### ISSUE THREAD

**Issue by rladydpf1** [diffblue/cbmc#7040]
CBMC version: 5.43.0
Operating system: Ubuntu 20.04.4 LTS
Exact command line resulting in the issue: cbmc --unwind 2 --slice-formula sv-benchmarks/c/combinations/gcd_3+newton_1_4.i
What behaviour did you expect: better performace than running CBMC without the '--slice-formula' option
What happened instead:
```
$ cbmc --unwind 2 --slice-formula sv-benchmarks/c/combinations/gcd_3+newton_1_4.i
...
58237 variables, 267895 clauses
SAT checker: instance is SATISFIABLE
Runtime Solver: 17.6473s
Runtime decision procedure: 17.7102s
```

```
$ cbmc --unwind 2 sv-benchmarks/c/combinations/gcd_3+newton_1_4.i
...
61129 variables, 280265 clauses
SAT checker: instance is SATISFIABLE
Runtime Solver: 9.34792s
Runtime decision procedure: 9.41339s
```
You can find the input file: https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/blob/main/c/combinations/gcd_3%2Bnewton_1_4.i
The '--slice-formula' option removed variables and clauses of SAT formula, but the runtime was bad.
Why did it happen?

**Comment by martin-cs** [comment]
Better performance is usual but not guaranteed.  There are two main reasons for this:

* The solver is applying heuristics.  It should be deterministic given the same formula and options but changing the formula changes the heuristic decisions.  Maybe the sliced one is an "unlucky" run, or maybe the unsliced one is a "lucky" formula.  This is part of why people do large scale benchmarks; to even out the influence of luck.
* There are cases where auxiliary formula can help improve the performance of the solver even though they do not obviously appear to be what you are proving.  The reasons for this are subtle and do not seem to be well studied.  Things like improving heuristics, giving a better vocabularly for learning, avoiding the need to learn certain implied facts, improving pre-processing or in-processing.

Basically, number of variables and number of clauses are both really poor metrics of the difficulty of a SAT problem but people like to use them.

What you have found is interesting but it is not a bug.  If it is something you really want to dig into I can try to help.

**Comment by rladydpf1** [comment]
Thank you for the answer.
